File: frontend/mobile_app/lib/features/customer/customer_directory_controller.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../../models/customer.dart';

/// A saved book of customer profiles the associate can build up over time.
///
/// The backend has no multi-profile endpoint yet, so the book is persisted
/// **locally** with [SharedPreferences]. Each entry is a full [Customer], so the
/// same model (and `toJson`) will POST unchanged once an API exists.
class CustomerDirectoryController extends ChangeNotifier {
  CustomerDirectoryController() {
    _restore();
  }

  static const String _kProfiles = 'customers.saved';

  List<Customer> _profiles = <Customer>[];
  bool _loaded = false;

  /// Saved profiles, most recently updated first.
  List<Customer> get profiles => List<Customer>.unmodifiable(_profiles);
  bool get isLoaded => _loaded;
  bool get isEmpty => _profiles.isEmpty;

  Future<void> _restore() async {
    try {
      final SharedPreferences prefs = await SharedPreferences.getInstance();
      final String? raw = prefs.getString(_kProfiles);
      if (raw != null && raw.isNotEmpty) {
        final List<dynamic> list = jsonDecode(raw) as List<dynamic>;
        _profiles = list
            .map((dynamic e) => Customer.fromJson(e as Map<String, dynamic>))
            .toList();
      }
    } catch (_) {
      _profiles = <Customer>[];
    }
    _loaded = true;
    notifyListeners();
  }

  Future<void> _persist() async {
    try {
      final SharedPreferences prefs = await SharedPreferences.getInstance();
      await prefs.setString(
        _kProfiles,
        jsonEncode(_profiles.map((Customer c) => c.toJson()).toList()),
      );
    } catch (_) {
      // Best-effort; the in-memory book still works for this session.
    }
  }

  /// Insert or update [customer] by id and persist. Returns the stored record
  /// (with a generated id when the draft didn't have a real one yet).
  Future<Customer> save(Customer customer) async {
    final bool needsId =
        customer.id.isEmpty || customer.id == 'draft';
    final Customer record = needsId
        ? customer.copyWith(id: 'local_${DateTime.now().microsecondsSinceEpoch}')
        : customer;

    final int index = _profiles.indexWhere((Customer c) => c.id == record.id);
    if (index >= 0) {
      _profiles[index] = record;
    } else {
      _profiles.insert(0, record);
    }
    notifyListeners();
    await _persist();
    return record;
  }

  Future<void> remove(String id) async {
    _profiles.removeWhere((Customer c) => c.id == id);
    notifyListeners();
    await _persist();
  }

  Customer? byId(String? id) {
    if (id == null) return null;
    for (final Customer c in _profiles) {
      if (c.id == id) return c;
    }
    return null;
  }

  /// A short human label for list rows.
  static String labelFor(Customer c) {
    final String name = (c.name ?? '').trim();
    if (name.isNotEmpty) return name;
    final String mobile = (c.mobile ?? '').trim();
    if (mobile.isNotEmpty) return mobile;
    return 'Unnamed guest';
  }

  /// A one-line summary of the guest's strongest preferences.
  static String summaryFor(Customer c) {
    final List<String> bits = <String>[
      ...c.fashionStyles.take(2),
      ?c.preferredFit,
      ?c.budgetRange,
      ...c.favoriteColors.take(2),
    ];
    return bits.isEmpty ? 'No preferences captured yet' : bits.join(' · ');
  }
}

File: frontend-rn/mobile_app/src/features/customer/customerDirectoryController.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Listenable } from '../../core/listenable';
import { Customer } from '../../models/customer';

/// A saved book of customer profiles persisted locally with AsyncStorage. Ported
/// 1:1 from `CustomerDirectoryController`.
export class CustomerDirectoryController extends Listenable {
  private static readonly K_PROFILES = 'customers.saved';

  private _profiles: Customer[] = [];
  private _loaded = false;
  private seq = 0;

  get profiles(): Customer[] {
    return this._profiles;
  }
  get isLoaded(): boolean {
    return this._loaded;
  }
  get isEmpty(): boolean {
    return this._profiles.length === 0;
  }

  constructor() {
    super();
    void this.restore();
  }

  private async restore(): Promise<void> {
    try {
      const raw = await AsyncStorage.getItem(CustomerDirectoryController.K_PROFILES);
      if (raw != null && raw.length > 0) {
        const list = JSON.parse(raw) as any[];
        this._profiles = list.map((e) => Customer.fromJson(e));
      }
    } catch {
      this._profiles = [];
    }
    this._loaded = true;
    this.notifyListeners();
  }

  private async persist(): Promise<void> {
    try {
      await AsyncStorage.setItem(
        CustomerDirectoryController.K_PROFILES,
        JSON.stringify(this._profiles.map((c) => c.toJson())),
      );
    } catch {
      // Best-effort.
    }
  }

  async save(customer: Customer): Promise<Customer> {
    const needsId = customer.id.length === 0 || customer.id === 'draft';
    const record = needsId ? customer.copyWith({ id: `local_${Date.now()}_${this.seq++}` }) : customer;
    const index = this._profiles.findIndex((c) => c.id === record.id);
    if (index >= 0) {
      this._profiles = this._profiles.map((c, i) => (i === index ? record : c));
    } else {
      this._profiles = [record, ...this._profiles];
    }
    this.notifyListeners();
    await this.persist();
    return record;
  }

  async remove(id: string): Promise<void> {
    this._profiles = this._profiles.filter((c) => c.id !== id);
    this.notifyListeners();
    await this.persist();
  }

  byId(id: string | null | undefined): Customer | null {
    if (id == null) return null;
    return this._profiles.find((c) => c.id === id) ?? null;
  }

  static labelFor(c: Customer): string {
    const name = (c.name ?? '').trim();
    if (name.length > 0) return name;
    const mobile = (c.mobile ?? '').trim();
    if (mobile.length > 0) return mobile;
    return 'Unnamed guest';
  }

  static summaryFor(c: Customer): string {
    const bits = [
      ...c.fashionStyles.slice(0, 2),
      c.preferredFit,
      c.budgetRange,
      ...c.favoriteColors.slice(0, 2),
    ].filter((x): x is string => x != null);
    return bits.length === 0 ? 'No preferences captured yet' : bits.join(' · ');
  }
}
